tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -164,357 +205,402 @@ public void invalidateBufferedData() throws IOException {
         // buffer from disk.
-        if (this.diskPos != this.lo) {
-            super.seek(this.lo);
-            this.diskPos = this.lo;
-        }
-        this.hi = this.lo + fillBuffer();
+        fillBuffer();
     }
-    
-    /* overrides RandomAccessFile.close() */
+
+    @Override
     public void close() throws IOException {
-        // Assert.check(!this.closed);
-        this.flush();
-        // SZ Feb 24, 2009: never read locally
-        // this.closed = true;
-        synchronized (mu) {
-            // grow "availBuffs" array if necessary
-            if (numAvailBuffs >= availBuffs.length) {
-                byte[][] newBuffs = new byte[numAvailBuffs + 10][];
-                System.arraycopy(availBuffs, 0, newBuffs, 0, numAvailBuffs);
-                availBuffs = newBuffs;
+        try {
+            // Assert.check(!this.closed);
+            this.flush();
+            // SZ Feb 24, 2009: never read locally
+            // this.closed = true;
+            synchronized (mu) {
+                // grow "availBuffs" array if necessary
+                if (numAvailBuffs >= availBuffs.length) {
+                    byte[][] newBuffs = new byte[numAvailBuffs + 10][];
+                    System.arraycopy(availBuffs, 0, newBuffs, 0, numAvailBuffs);
+                    availBuffs = newBuffs;
+                }
+                availBuffs[numAvailBuffs++] = this.buff;
             }
-            availBuffs[numAvailBuffs++] = this.buff;
+        } finally {
+            super.close();
         }
-        super.close();
     }
-    
-    /** Flush any bytes in the file's buffer that have not
-        yet been written to disk. If the file was created
-        read-only, this method is a no-op. */
+
+    /**
+     * Flush any bytes in the file's buffer that have not
+     * yet been written to disk. If the file was created
+     * read-only, this method is a no-op.
+     */
     public void flush() throws IOException {
         // Assert.check(!this.closed);
         this.flushBuffer();
     }
-    
-    /* Flush any dirty bytes in the buffer to disk. */
+
+    /**
+     * Flush any dirty bytes in the buffer to disk.  Sets {@link #dirty} to <code>false</code>.
+     *
+     * @return true iff any disk writes were performed
+     */
     private boolean flushBuffer() throws IOException {
         if (this.dirty) {
             // Assert.check(this.curr > this.lo);
-            if (this.diskPos != this.lo) super.seek(this.lo);
-            int len = (int)(this.curr - this.lo);
-            super.write(this.buff, 0, len); 
-            this.diskPos = this.curr;
+            int len = (int)Math.min(this.length - this.lo, BuffSz);
+            if (len > 0) {
+                assert super.getFilePointer() == diskPos;
+                if (this.diskPos != this.lo) {
+                    super.seek(this.lo);
+                }
+                super.write(this.buff, 0, len);
```

Comment:
It is weird, but [`write` takes an int as its third argument](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/RandomAccessFile.html#write(byte%5B%5D,int,int)). That is because Java arrays can only contain an `int`'s worth of entries, and the `len` argument specifies how many entries to read from the `buff` array.

In this case `len` is guaranteed to fit in an `int`, since it can't be larger than `BuffSz` and `BuffSz` is an `int`.